test/lsp/cache_protocol_test_corpus.cc

```diff
@@ -258,4 +259,72 @@ TEST_CASE_FIXTURE(CacheProtocolTest, "LSPDoesNotUseCacheIfModified") {
         }
     }
 }
+
+TEST_CASE_FIXTURE(CacheProtocolTest, "ReindexingUsesTheCache") {
+    // Write a file to disk.
+    auto relativeFilepath = "test.rb";
+    auto filePath = fmt::format("{}/{}", rootPath, relativeFilepath);
+    // This file has an error to indirectly assert that LSP is actually typechecking the file during initialization.
+    auto fileContents = "# typed: true\n"
+                        "class Foo extend T::Sig\n"
+                        "  sig {returns(Integer)}\n"
+                        "  def bar\n"
+                        "    'hello'\n"
+                        "  end\n"
+                        "end\n";
+    auto key = core::serialize::Serializer::fileKey(
+        core::File(string(filePath), string(fileContents), core::File::Type::Normal, 0));
+
+    // LSP should write a cache to disk corresponding to initialization state.
+    {
+        writeFilesToFS({{relativeFilepath, fileContents}});
+
+        lspWrapper->opts->inputFileNames.push_back(filePath);
+        assertErrorDiagnostics(
+            initializeLSP(),
+            {{relativeFilepath, 4, "Expected `Integer` but found `String(\"hello\")` for method result type"}});
+    }
+
+    // LSP should have written cache to disk with file hashes from initialization.
+    // It should not include data from file updates made during the editor session.
+    auto opts = lspWrapper->opts;
+
+    // Release cache lock.
+    lspWrapper = nullptr;
```

Comment:
Do we create an `lspWrapper` for every test? Is this state ever shared across tests in a way that would make our test suite flaky?

Reply:
I'm not entirely sure. I can read in more detail, but my expectation was that the existing test corpus would be exhibiting flakyness if that were true.

Reply:
Yeah afterwards I realized that there are other tests that set `lspWrapper = nullptr` like this.

Reply:
We do create an `lspWrapper` for each test, but the `CacheProtocolTest` inherits from `ProtocolTest`, which goes to great lengths to sandbox the filesystem, and create a tempdir for the cache on each instantiation. I think this shouldn't cause any flakyness.

Reply:
> Yeah afterwards I realized that there are other tests that set `lspWrapper = nullptr` like this.

Might be worth adding `this->` for those to make it obvious that it's an instance variable.